Instancer : Support relative prototype paths

danieldresser-ie opened the pull request:
I'm deleting the previous discussion here, and starting over, because this is a completely different PR to the first time around.

It is now built on top of #6258, #6272, and #6271, ( in addition to requiring the cortex change https://github.com/ImageEngine/cortex/pull/1451 ). Assuming you want to review this in one go, those other PRs can be closed.

I'm feeling fairly good about this - seems to do what we want, with fairly decent test coverage. This includes performance tests - the one noticeable regression is testPrototypeHashPerf, which went from 0.02s to 0.13s. This test was written specifically to demonstrate worst case behaviour - in practice, we usually need to actually evaluate capsules, not just hash them, and the gain in consistency and performance from reusing capsules when possible should outweigh the cost of the much more accurate hash.

While I'm feeling fairly good, this is a significant change to some code that interacts in fairly complicated ways, so it does call for careful review. If you have access to production data from the users requesting the feature, it would be great to confirm that it works as desired in practice.

The only thing I know still needs doing is writing changelog entries - since this isn't getting merged this week, I'm expecting to need to do some rebasing, so I'll write changelog entries afterwards.

danieldresser-ie commented:
Note that there are no tests here yet - I figured we'd figure out what the behaviour we want is first, and then I'll write some tests for that. ( In interactive testing, this does seem to accomplish the goal of making it possible to relocate USD pointInstancers and having everything work properly ).

johnhaddon reviewed (commented):
Sorry Daniel, domestic duties beckon and I haven't quite finished the review on this - need to give a bit more attention to the relative paths commit in particular. Figured it was worth posting what I had in the meantime though.

I pushed a couple of commits on the end to update Cortex and update the env-var name to match, but now https://github.com/GafferHQ/gaffer/pull/6287 is merged you should be able to rebase and drop the Cortex commit.

Not sure why `testRelativePrototypePaths` is failing on Linux but not Windows. It is passing for me locally on Linux. Maybe it's obvious to you why?

Cheers...
John

danieldresser-ie commented on include/GafferScene/Instancer.h:
```diff
@@ -245,6 +245,9 @@ class GAFFERSCENE_API Instancer : public BranchCreator
 		Gaffer::PathMatcherDataPlug *setCollaboratePlug();
 		const Gaffer::PathMatcherDataPlug *setCollaboratePlug() const;
 
+		Gaffer::Int64VectorDataPlug *capsuleComputedHashPlug();
+		const Gaffer::Int64VectorDataPlug *capsuleComputedHashPlug() const;
```
This would be easy if we were actually only using the hash for constructing the capsule, but currently, we're using the accurate hash as the actual hash as well.

I can think of two possible performance sacrifices we could make in order to simplify this:
A) Don't use the accurate hash in Instancer::hashObject. Then we can use your capsulePlug() idea. The cost of this is that now moving something unrelated in the prototypes scene invalidates the object in the value cache. This would trigger extra computes in scenarios like moving a light when using the USD adaptor ( though once we hand the capsule to the render backend, it would be able to see that actual hash is the same, so it can still instance things properly ).
B) Compute the accurate hash in the hash of a new plug, instead of the compute ( would that make it less funky? ). The cost of this is that moving something unrelated in the input scene would force the recomputation of the expensive hash. Worst case for this is that your using an explicit instancer, not a USD adaptor, so the input and prototype scene are different, and the prototype scene is incredibly heavy. Currently, if you just move a light in the input scene, the expensive hash is just pulled from the compute, but in this scheme you'd have to do a complete hierarchyHash of all the prototypes.

danieldresser-ie replied:
After talking through this, we didn't come up with anything better than what we're currently doing. We did come up with option C) have a plug that stores the computed capsule, and have hashObject use the accurate from that capsule, which seems quite attractive ... the problem is that creating the Capsule includes an expensive bounds calculation, and we wouldn't want hashObject to depend on that, so we would actually need a plug that stores "capsules but with their bounds not set yet", and Instancer::computeObject would have to get the capsule from the capsulesWithoutBoundsPlug(), and then set the bound before returning it. At this point, option C is kinda funky too, and maybe it's just as well to stick with what we've got here.

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -2460,25 +2606,28 @@ void Instancer::hashObject( const ScenePath &path, const Gaffer::Context *contex
-			}
-			engineHash( sourcePath, context, h );
-			h.append( context->hash() );
-			outPlug()->boundPlug()->hash( h );
```
Oh, good point, there was an intermediate commit where I was thinking about this, but missed the comment during cleanup: we currently include the hash of the engine, and the hash of all the prototype bounds. This should include everything that affects the bound already ... though it's true that this doesn't make it easy to match to the compute method, which does use outPlug()->boundPlug. The best solution from a performance point of view would probably be to implement the bound computation ourselves ( doing the accumulation ourselves using the list of prototypes and the prototype indices would be more efficient than using the generic childBounds computation ), but I haven't bothered doing this yet.

johnhaddon commented on src/GafferScene/Instancer.cpp:
```diff
@@ -2989,6 +3065,18 @@ void Instancer::PrototypeScope::setPrototype( const EngineData *engine, const Sc
 		m_prototypePath.insert( m_prototypePath.end(), branchPath->begin() + 3, branchPath->end() );
 		set( ScenePlug::scenePathContextName, &m_prototypePath );
 	}
+	else if( prototypeRootStorage.size() )
+	{
+		// Slightly awkward case here: prototypeRoot may need to allocate storage for the prototype root path.
+		// ( This happens when it's a relative path, and we need to prepend the source path ).
+		// If we had to allocate the path, we need to make sure we store it, rather than leaving a dangling
+		// pointer.
+		//
+		// If we didn't need to allocate, the engine will hold the memory for the path, and we don't want to
+		// copy it, so we can take the branch below.
+		m_prototypePath = prototypeRootStorage;
+		set( ScenePlug::scenePathContextName, &m_prototypePath );
+	}
```
I wonder if it's less awkward to pass `m_prototypePath` for the storage in the first place? We'd need to adjust the `branchPath->size() > 3` block slightly, but overall I think it might be simpler and would avoid some copies?

As an aside, I don't know what the `reserve()` is doing in that block - surely `insert()` is optimised for cases where it can get the distance between iterators in constant time?

danieldresser-ie replied:
Lets see, if we just always allocated the path this could definitely get a bit simpler ( I don't even think we'd need to pass m_prototypePath in, we could just return a ScenePath from prototypeRoot, instead of a pointer ). And it would save an allocation when you're on a child of the prototype, and the prototype is relative.

But my concern is that the special case where no allocation is currently needed could easily be a common case: prototype paths that aren't relative, and prototypes with no children. Are you happy to switch to doing an allocation in this case? Should I write a perf test specifically to see if there's any measurable performance loss in the worst case?

On your aside: technically, the reserve is actually improving the allocation behaviour here: we know that we are just resizing m_prototypePath once, and we know exactly the final size. Just doing an `insert` would grow the capacity more than needed, by some multiplicative factor, in order to amortize the cost of a sequence of repeated inserts, which is unnecessary here because there's only one insert.

danieldresser-ie replied:
I realized you were actually just asking for quite a minor refactor which only really affects a couple branch of this call site, I've done that here: 444a94b60d18d14adb3e1c7ba0ad1f2d7519d52b

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -2588,6 +2584,12 @@ IECore::ConstInternedStringVectorDataPtr Instancer::computeBranchChildNames( con
 		ids.reserve( pointIndicesForPrototype.size() );
 
 		const EngineData *engineData = esp->engine();
+		const ScenePlug::ScenePath *prototypeRoot = engineData->prototypeRoot( branchPath[1] );
```
Addressed in 6fc691ff08f3a6119fd8b03770a29d7d7fa4a417 - it doesn't match the hash exactly, since I didn't really want to do a separate compute of enginePlug after already getting EngineSplitPrototypesData with a pointer to EngineData, but there's already a form of PrototypeScope that takes an existing EngineData pointer, so I can use that.

danieldresser-ie commented on bin/__gaffer.py:
```diff
@@ -68,6 +68,11 @@
 		resource.setrlimit( resource.RLIMIT_NOFILE, ( hardFileLimit, hardFileLimit ) )
 		IECore.msg( IECore.Msg.Level.Debug, "Gaffer", "Increased file handle limit to {}".format( hardFileLimit ) )
 
+# Set this env var flag to enable relative prototypes paths output from the IECoreUSD point instancer loader,
```
Looks like you've addressed this in your fix where you corrected the env var name.

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -387,7 +387,6 @@ class Instancer::EngineData : public Data
 			const std::string &prototypeIndexName,
 			const std::string &rootsVariable,
 			const StringVectorData *rootsList,
-			const ScenePlug *prototypes,
```
Squashed.

danieldresser-ie commented on python/GafferSceneTest/InstancerTest.py:
```diff
@@ -1307,7 +1307,7 @@ def testIndexedRootsList( self ) :
 		self.assertRaisesRegex(
 			Gaffer.ProcessException, '.*Prototype root "/does/not/exist" does not exist.*',
-			script["instancer"]["out"].childNames, "/object/instances",
+			script["instancer"]["out"].childNames, "/object/instances/exist",
```
Squashed.

danieldresser-ie commented on python/GafferSceneTest/InstancerTest.py:
```diff
@@ -3481,5 +3481,127 @@ def testEncapsulatedRenderPerf( self ):
+		# It should be very cheap to retrieve the cached capsule since we've only changed something irrelevant
+		with GafferTest.TestRunner.PerformanceScope() :
+
+			instancer["out"].object( "/plane/instances" )
```
Addressed in c5cd54327ea4afe87e0433204fefc21f388fb6a0

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -2143,6 +2180,111 @@ void Instancer::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
+		}
+		else
+		{
+			std::atomic<uint64_t> h1Accum( 0 ), h2Accum( 0 );
```
I tried measuring while working on hierarchyHash, and couldn't observe a difference there. Since this calls hierarchyHash for each iteration, it seems even less likely that these iterations could be fast enough for there to be an observable difference. After implementing this using parallel_reduce, I'm not sure it really feels clearer, but I'm fine with it if you prefer it: a8a4144d7221193b4b1539414ec79f8809a9b4fb

johnhaddon replied:
I'm not sure why we need the OrderIndependentHashAccum class for this - I've pushed a couple of commits at the end that use bare MurmurHashes, for your consideration.

The reason I like `parallel_reduce()` over `parallel_for()` in this case is that it's immediately clear that we're munging a bunch of stuff into a single value.

johnhaddon replied:
Note : even if you decide to omit my commits, they do contain the remove of some otherwise unused variables, so we should at least keep that part.

danieldresser-ie replied:
I'm a bit dubious of using operator+ here for MurmurHashes ... I feel like summing hashes is a rather unusual thing to do, and if we did want to make it part of the MurmurHash interface, it would kind of call for a more specific function name.

Rather than worry about that currently, I've gone with your first option ( where you explicitly wrote out all the adds, instead of using a new operator+ ). I've made two changes on top of this: I explicitly initialize the identity MurmurHashes to `( 0, 0 )` ( this is the default initialization, but that isn't documented ... for most usage, it would be fine if there was some other default, but when using them as accumulators, we require the identity to be 0 ). And I store the results of the two reduces in separate variables, and add them both to the result hash, instead of summing them ( was just a bit easier this way, and we don't need/want order independence between sets and locations ).

I'm assuming those 2 minors changes are acceptable, so I've gone ahead and squashed everything.

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -2143,6 +2180,111 @@ void Instancer::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
+			h.append( IECore::MurmurHash( h1Accum, h2Accum ) );
+		}
+
+		Int64VectorDataPtr resultData = new Int64VectorData();
```
Addressed in a6a3a2c373ee505a06ddbe807e7561df6179bb0a

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -1037,30 +1064,56 @@ class Instancer::EngineData : public Data
 			m_prototypeIndexRemap.reserve( rootStrings->size() );
 
+
+			const static bool gafferSceneInstancerExplicitAbsolutePaths = checkEnvFlag( "GAFFERSCENE_INSTANCER_EXPLICITABSOLUTEPATHS", false );
```
Addressed in 5c9b1d9e09d68b5183872077589617bd7fbf14fa

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -1075,7 +1128,13 @@ class Instancer::EngineData : public Data
 		size_t m_numValidPrototypes;
 		Private::ChildNamesMapPtr m_names;
-		std::vector<ConstInternedStringVectorDataPtr> m_roots;
+		struct ProtoRoot {
```
Addressed in 5c9b1d9e09d68b5183872077589617bd7fbf14fa

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -1075,7 +1128,13 @@ class Instancer::EngineData : public Data
+		struct ProtoRoot {
+			ProtoRoot( bool rel, ConstInternedStringVectorDataPtr path ) : rel( rel ), path( path ) {};
+
+			bool rel;
```
Addressed in 5c9b1d9e09d68b5183872077589617bd7fbf14fa

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -1037,30 +1064,56 @@ class Instancer::EngineData : public Data
+				{
+					m_prototypeIndexRemap.emplace_back( -1 );
+					continue;
+				}
```
Addressed in 5c9b1d9e09d68b5183872077589617bd7fbf14fa

danieldresser-ie commented:
The switch to using node ids instead of addresses appears to have addressed the test failures.

Once you've looked through my replies to comments, we just need to figure out what to do with prototypeRootStorage, and then I'll need to squash it, and sort out Changes entries.

danieldresser-ie commented:
I think all feedback is now addressed, and I've added some changelog entries - in theory, all that's left is squashing.

johnhaddon commented:
> I think all feedback is now addressed, and I've added some changelog entries - in theory, all that's left is squashing.

Thanks! I pushed a couple of slight tweaks for your consideration, but otherwise all looks ready to squash and merge.

We also need to fix the broken layout for the time offset in the Context Variation tab - could you either do that in this PR or open a separate PR for it please?

danieldresser-ie commented:
Squashed in a version of your tweak ( see comment above ), and fixed the broken layout. I understand most of why the layout broke: it was mixing custom widgets with explicit layout indices with plugs with default layout indices, so changing the indices of the plugs caused everything to get offset ( I've addressed this by assigning explicit layout indices to everything on that tab, which should prevent the problem in the future ).

But the part I don't get is why the alignment of the column labels broke ... or more specifically I don't get how it worked previously. We were specifying `horizontalAlignment=GafferUI.HorizontalAlignment.Left` on the label ... but it looks like the label was fixed size anyway, so the alignment wasn't doing much? I've now added an extra spacer, which seems to have gotten the column alignment back the way it should be ... just feeling a bit uncomfortable that I'm not confident why this was needed.

johnhaddon commented:
> I've now added an extra spacer, which seems to have gotten the column alignment back the way it should be ... just feeling a bit uncomfortable that I'm not confident why this was needed.

I haven't verified, but I assume this will have been down to bb1f4327d6cab26de4f2a1b1770d42f08f402447. All good I think.
